src/Security/InheritedPermissions.php

```diff
@@ -0,0 +1,632 @@
+            );
+            if ($actuallyInherited) {
+                $parentIDs = array_keys(array_filter($actuallyInherited));
+                foreach ($parentIDs as $parentID) {
+                    // Set all the relevant items in $result to true
+                    $result = array_fill_keys($groupedByParent[$parentID], true) + $result;
+                }
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Get the 'can edit' information for a number of SiteTree pages.
+     *
+     * @param array $ids An array of IDs of the objects to look up
+     * @param Member $member Member object
+     * @param bool $useCached Return values from the permission cache if they exist
+     * @return array A map where the IDs are keys and the values are
+     * booleans stating whether the given object can be edited
+     */
+    public function canEditMultiple($ids, Member $member = null, $useCached = true)
+    {
+        return $this->batchPermissionCheck(
+            self::EDIT,
+            $ids,
+            $member,
+            $this->getGlobalEditPermissions(),
+            $useCached
+        );
+    }
+
+    /**
+     * Get the canView information for a number of objects
+     *
+     * @param array $ids
+     * @param Member $member
+     * @param bool $useCached
+     * @return mixed
+     */
+    public function canViewMultiple($ids, Member $member = null, $useCached = true)
+    {
+        return $this->batchPermissionCheck(self::VIEW, $ids, $member, [], $useCached);
+    }
+
+    /**
+     * Get the 'can edit' information for a number of SiteTree pages.
+     *
+     * @param array $ids An array of IDs of the objects pages to look up
+     * @param Member $member Member object
+     * @param bool $useCached Return values from the permission cache if they exist
+     * @return array
+     */
+    public function canDeleteMultiple($ids, Member $member = null, $useCached = true)
+    {
+        // Validate ids
+        $ids = array_filter($ids, 'is_numeric');
+        if (empty($ids)) {
+            return [];
+        }
+        $result = array_fill_keys($ids, false);
+
+        // Validate member permission
+        if (!$member || !$member->ID) {
+            return $result;
+        }
+        $deletable = [];
+
+        // Look in the cache for values
+        $cacheKey = "delete-{$member->ID}";
+        if ($useCached && isset($this->cachePermissions[$cacheKey])) {
+            $cachedValues = array_intersect_key($this->cachePermissions[$cacheKey], $result);
+
+            // If we can't find everything in the cache, then look up the remainder separately
+            $uncachedIDs = array_keys(array_diff_key($result, $this->cachePermissions[$cacheKey]));
+            if ($uncachedIDs) {
+                $uncachedValues = $this->canDeleteMultiple($uncachedIDs, $member, false);
+                return $cachedValues + $uncachedValues;
+            }
+            return $cachedValues;
+        }
```

Comment:
The code blocks are similar but not quite the same; There are some subtle differences that would make refactoring less useful for de-duplication. In this case, it was simpler and more readable to simply repeat the pattern.